### ISSUE TITLE
[REQ] Tactile Feedback for Zwift Play controllers (and Ride?)

### ISSUE THREAD

**Issue by GraVvVel** [cagnulein/qdomyos-zwift#2752]
**Is your feature request related to a problem? Please describe.**
Nope

**Describe the solution you'd like**
On Zwift gearing you get subtle tactile feedback in the Zwift Play controllers for each gear shift. I would be nice to replicate this in QZ when connecting Zwift Play controllers with it and using it for gear shifts 

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by cagnulein** [comment]
Thanks, I'm already writing the code

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 7 nov 2024 alle 13:00 GraVvVel ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
> So are you saying it won't be possible to implement the vibration? Would be happy to do any testing, if it helps.

no i just said that i tried but it's harder than i was expecting. I just gave it a single test. If a ding is not useful let's stay with the master plan :)

I will try again in the next weeks

**Comment by cagnulein** [comment]
TTS is it already a thing, if you want i can add a tts for the gear too

**Comment by satchafunkilus** [comment]
I know about the TTS, but as far as I understand only reads the current values in certain predefined intervals. In this case we would need a read out directly when shifting, or ideally with a small delay (1s?), so that it ignores multiple quick shifts at once and only reads out the "final" gear

**Comment by cagnulein** [comment]
yes i mean I can do it :) but i guess for you the vibrate feedback is better or also tts in tandem with it could be a good thing?

**Comment by satchafunkilus** [comment]
It's two completely different topics, I would say. I would love both.

**Comment by cagnulein** [comment]
@satchafunkilus are you watching this too? https://github.com/cagnulein/qdomyos-zwift/issues/2751#issuecomment-2504621022
I just finished it in case you are interested

**Comment by satchafunkilus** [comment]
Sounds cool, happy to check it out as well. For this function in particular, I could imagine a beep/click or some audible feedback to make sense, so you know how many gears have changed, especially if there is a small delay in the trainer. Could of course also be solved by the vibration ideally, but I think in this case some fallback would be great, as long as we don't have vibration. Otherwise I would imagine it very hard to know how long to keep the button pressed.

**Comment by cagnulein** [comment]
@satchafunkilus 
send me an email to roberto.viola83 at gmail.com indicating this ticket and if you are on android or iOS. Thanks

i added a 0.5 delay. so it's backward compatible, as long as you are pushing more it's increasing too. it feels very similar to what zwift does i guess. let me know!

**Comment by VictorSharov** [comment]
Yes, think the same. Will be great to have a tactile feedback about gear shifting. Since the buttons are a bit stiff, sometimes I don't understand whether is press registered or not. Tactile feedback on Zwift controllers will be more comfort and better. But vibration (haptic feedback) on phone is better than nothing.
What's so hard about getting a vibration command from Zwift? They encrypt the data exchange and it's hard to figure out what command they use for vibration?

**Comment by cagnulein** [comment]
good news! handled!

**Comment by satchafunkilus** [comment]
Just gave it a quick spin. It only vibrated on some shifts (maybe 10-20% of the time) and only with a pretty big delay of a second or so. 

[debug-Sa__Dez__28_13_31_58_2024.log](https://github.com/user-attachments/files/18266163/debug-Sa__Dez__28_13_31_58_2024.log)

**Comment by satchafunkilus** [comment]
Appears to be linked to the debouncing. When I turn off debouncing, it vibrates every time, but then also shifts 3-4 gears at once

**Comment by cagnulein** [comment]
@satchafunkilus I'm checking thanks! Strange! With my zwift play it works fine! But for sure there is a explanation!

**Comment by cagnulein** [comment]
@satchafunkilus Is it possible that it works only when you are going down with the gears and not up?

**Comment by cagnulein** [comment]
I guess I got it, I'm attaching a screen for me as a note
![image](https://github.com/user-attachments/assets/253e8fca-3cc4-4933-b795-e1269b6d2e7b)

@satchafunkilus Just confirm me that it's working only one way. If it's so, I guess I got it

**Comment by satchafunkilus** [comment]
> @satchafunkilus Is it possible that it works only when you are going down with the gears and not up?

Yes, currently it’s working quite consistently. It vibrates twice on downshifts and not on upshifts.

**Comment by cagnulein** [comment]
Perfect I got, I will send you a new one later today 

I tested with zwift play and they have a small little difference

**Comment by cagnulein** [comment]
@satchafunkilus new one sent, let me know!

**Comment by satchafunkilus** [comment]
Fix works great now on zwift ride. One suggestion I have is to give a different feedback when you have reached the end of your gears and it cannot shift any further. So if you‘re on gear 1 and try to shift down it could do a triple vibration or a long vibration or something like that. If this is not possible, we could maybe have no vibration in those cases.

**Comment by cagnulein** [comment]
mmm ok i will think about it (because now, in order to send it without latency, i'm doing any control at all). i will play around with it.

**Comment by cagnulein** [comment]
done @satchafunkilus i'm sending you a new one!

**Comment by satchafunkilus** [comment]
This is really awesome now. Thanks so much 👌

**Comment by cagnulein** [comment]
@satchafunkilus Amazing! Did you try also the new build with the settings improved? I improved the speed of the settings menu and also I binded the fact that the switch button can switch only if you click on the switch and not to the whole line

**Comment by satchafunkilus** [comment]
Yes I did - works fine 👍
